### PR TITLE
Feature: Recovery

### PR DESCRIPTION
**Pull request by maxplt**
# Jira Issue: [IP-281](https://hedvig.atlassian.net/jira/software/projects/IP/boards/19?selectedIssue=IP-281)

## What?
Add recovery

## Why?
-

## Optional screenshots
image

## Optional checklist
- [ ] Updated `src/changelog.ts`
- [x] Codescouted
- [ ] Unit tests written
- [ ] Tested locally

### CODE REVIEW

**Comment by falklinus** [comment]
image
image

File: src/portals/hope/features/claims/claim-details/ClaimCounterparty/CounterpartyDetails.tsx [hunk 1]
```diff
@@ -0,0 +1,186 @@
...
+  return (
+    <>
+      <Fields>
+        <Input
+          disabled={true}
+          label="External Reference"
+          value={counterparty.reference}
+        />
+        <Select
+          label="Insurance Company (if applicable)"
```
Comment (vonElfvin) [inline src/portals/hope/features/claims/claim-details/ClaimCounterparty/CounterpartyDetails.tsx:94]:
```suggestion
          label="Insurance Provider (if applicable)"
```

File: src/portals/hope/features/claims/claim-details/ClaimCounterparty/CounterpartyDetails.tsx [hunk 2]
```diff
@@ -0,0 +1,186 @@
...
+          Show Invoice
+        </Button>
+      ) : (
+        <Button
+          type="submit"
+          onClick={() =>
+            createRecoveryInvoice({
+              variables: {
+                counterpartyId: counterparty.id,
+                memberReference: 'XXX999',
```
Comment (vonElfvin) [inline src/portals/hope/features/claims/claim-details/ClaimCounterparty/CounterpartyDetails.tsx:167]: This we should fetch from the registrationNumber of the `agreement` of the claim

Comment (vonElfvin) [reply src/portals/hope/features/claims/claim-details/ClaimCounterparty/CounterpartyDetails.tsx:167]: Also thinking about it we should probably not even show recovery if there is no `registrationNumber` on the agreemnet of the claim

File: src/portals/hope/features/claims/claim-details/ClaimCounterparty/CreateCounterpartyModal.tsx [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+      </Dropdown>
+      <Input
+        label="External reference"
+        placeholder={
+          counterpartyType === CounterpartyType.Car
+            ? 'Registration number'
+            : 'Company name, etc...'
+        }
+        value={reference}
+        onChange={(e) => {
```
Comment (vonElfvin) [inline src/portals/hope/features/claims/claim-details/ClaimCounterparty/CreateCounterpartyModal.tsx:104]: I know I wrote this but 🤪 
```suggestion
        onChange={({currentTarget: {value}}) => {
```

File: src/portals/hope/features/claims/claim-details/ClaimCounterparty/InvoiceModal.tsx [hunk 4]
```diff
@@ -0,0 +1,218 @@
...
+      paymentDueDate
+      counterparty {
+        id
+        reference
+        insuranceProvider {
+          displayName
+        }
+      }
+    }
+  }
```
Comment (vonElfvin) [inline src/portals/hope/features/claims/claim-details/ClaimCounterparty/InvoiceModal.tsx:69]: Pretty sure the only thing that would update is the content
```suggestion
    ) {
      id
      content
    }
  }
```

File: src/portals/hope/features/claims/claim-details/ClaimCounterparty/InvoiceModal.tsx [hunk 5]
```diff
@@ -0,0 +1,218 @@
...
+`
+
+export const InvoiceModal: React.FC<{
+  visible: boolean
+  onClose: () => void
+  invoice: RecoveryInvoice | null
```
Comment (vonElfvin) [inline src/portals/hope/features/claims/claim-details/ClaimCounterparty/InvoiceModal.tsx:75]: Shouldn't be possible for this to be null
```suggestion
  invoice: RecoveryInvoice
```

File: src/portals/hope/features/claims/claim-details/ClaimCounterparty/InvoiceModal.tsx [hunk 6]
```diff
@@ -0,0 +1,218 @@
...
+}> = ({ visible, onClose, invoice }) => {
+  const [updateRecoveryInvoiceContent] =
+    useUpdateRecoveryInvoiceContentMutation()
+
+  const content =
+    invoice === null
+      ? ''
+      : `
```
Comment (vonElfvin) [inline src/portals/hope/features/claims/claim-details/ClaimCounterparty/InvoiceModal.tsx:83]: With the change above
```suggestion
  const content = `
```

File: src/portals/hope/features/claims/claim-details/ClaimCounterparty/InvoiceModal.tsx [hunk 7]
```diff
@@ -0,0 +1,218 @@
...
+          <p style="margin:0">
+            The amount can be deposited in our Bank Giro: 
+            <span style="color:${lightTheme.accent}">${
+          invoice.bankgiroNumber
+        }</span> 
+            or Bank Account: 
+            <span style="color:${lightTheme.accent}">${
+          invoice.bankAccountNumber
+        }</span>
+          </p>
```
Comment (vonElfvin) [inline src/portals/hope/features/claims/claim-details/ClaimCounterparty/InvoiceModal.tsx:156]: I think it'd be easier to read these if they were on separate lines
```suggestion
          <p style="margin:0">
            The amount can be deposited in our:</p>
         <p style="margin:0">
            Bank Giro: 
            <span style="color:${lightTheme.accent}">${
          invoice.bankgiroNumber
        }</p> 
        <p style="margin:0">
            Bank Account: 
            <span style="color:${lightTheme.accent}">${
          invoice.bankAccountNumber
        }</span>
          </p>
```

File: src/portals/hope/features/claims/claim-details/ClaimCounterparty/InvoiceModal.tsx [hunk 8]
```diff
@@ -0,0 +1,218 @@
...
+          <p style="margin:0">
+            Payment must be received by us no later than 
+            <span style="color:${lightTheme.accent}">${
+          invoice.paymentDueDate
+        }</span>.
+          </p>
+        </div>`
+
+  return (
+    <InvoiceModalStyled visible={visible} onClose={onClose}>
```
Comment (vonElfvin) [inline src/portals/hope/features/claims/claim-details/ClaimCounterparty/InvoiceModal.tsx:172]:
```suggestion
    <InvoiceModalStyled visible={true} onClose={onClose}>
```

Comment (vonElfvin) [reply src/portals/hope/features/claims/claim-details/ClaimCounterparty/InvoiceModal.tsx:172]: I think we can skip the `visible` property above as well

File: src/portals/hope/features/claims/claim-details/ClaimCounterparty/InvoiceModal.tsx [hunk 9]
```diff
@@ -0,0 +1,218 @@
...
+      <ThirdLevelHeadline>Invoice Preview</ThirdLevelHeadline>
+      <Remark>
+        All recovery records of the selected party with status ‘Pending Invoice’
+        will be included.
+        <br />
+        Invoice number will be generated once you confirm the content by
+        clicking ‘Confirm’
```
Comment (vonElfvin) [inline src/portals/hope/features/claims/claim-details/ClaimCounterparty/InvoiceModal.tsx:179]: This is actually not true, the invoice is created when pressing the buttom. We should maybe move this into a `confirm` when pressing `Generate invoice` instead. Also the copy would be:
"All recovery records of the selected counterparty will be included"

File: src/portals/hope/features/claims/claim-details/ClaimCounterparty/InvoiceModal.tsx [hunk 10]
```diff
@@ -0,0 +1,218 @@
...
+      <Spacing top="small" />
+      <ButtonsGroup>
+        {invoice !== null && (
+          <Button
+            onClick={() =>
+              toast.promise(
+                updateRecoveryInvoiceContent({
+                  variables: {
+                    recoveryInvoiceId: invoice.id,
+                    content: content ?? '',
```
Comment (vonElfvin) [inline src/portals/hope/features/claims/claim-details/ClaimCounterparty/InvoiceModal.tsx:194]: `content` should always exist right?

File: src/portals/hope/features/claims/claim-details/ClaimCounterparty/InvoiceModal.tsx [hunk 11]
```diff
@@ -0,0 +1,218 @@
...
+                  },
+                }),
+                {
+                  loading: 'Loading',
```
Comment (vonElfvin) [inline src/portals/hope/features/claims/claim-details/ClaimCounterparty/InvoiceModal.tsx:198]:
```suggestion
                  loading: 'Updating recovery invoice content...',
```

File: src/portals/hope/features/claims/claim-details/ClaimCounterparty/InvoiceModal.tsx [hunk 12]
```diff
@@ -0,0 +1,218 @@
...
+                    copy(content, { format: 'text/html' })
+                    onClose()
+                    return 'Content copied to clipboard'
+                  },
+                  error: 'Could not set content',
+                },
+              )
+            }
+          >
+            Confirm and copy to clipboard
```
Comment (vonElfvin) [inline src/portals/hope/features/claims/claim-details/ClaimCounterparty/InvoiceModal.tsx:209]: Neat!

File: src/portals/hope/features/claims/claim-details/ClaimCounterparty/index.tsx [hunk 13]
```diff
@@ -0,0 +1,123 @@
...
+  const counterparties = data?.claim?.counterparties ?? []
+
+  useEffect(() => {
+    if (counterparties.length === 0) return
+    if (selectedCounterparty === null) {
+      setSelectedCounterparty(counterparties[0])
+      return
+    }
+    setSelectedCounterparty(
+      counterparties.find((cp) => cp.id === selectedCounterparty.id) ?? null,
```
Comment (vonElfvin) [inline src/portals/hope/features/claims/claim-details/ClaimCounterparty/index.tsx:89]:
```suggestion
      counterparties.find((counterparty) => counterparty.id === selectedCounterparty.id) ?? null,
```

File: src/portals/hope/features/claims/claim-details/ClaimCounterparty/index.tsx [hunk 14]
```diff
@@ -0,0 +1,123 @@
...
+    )
+  }, [counterparties])
```
Comment (vonElfvin) [inline src/portals/hope/features/claims/claim-details/ClaimCounterparty/index.tsx:91]:
```suggestion
  }, [counterparties.length])
```

File: src/portals/hope/features/claims/claim-details/ClaimSubclaims/ClaimSubclaims.tsx [hunk 15]
```diff
@@ -90,7 +99,8 @@ const Subclaim: React.FC<{
   subclaim: ClaimInformationSubclaimFragment
   memberId: string
   claimId: string
-}> = ({ subclaim, memberId, claimId }) => {
+  state?: ClaimState
```
Comment (vonElfvin) [inline src/portals/hope/features/claims/claim-details/ClaimSubclaims/ClaimSubclaims.tsx:102]: ClaimState should always exist on a claim
```suggestion
  state: ClaimState
```

File: src/portals/hope/features/claims/claim-details/ClaimSubclaims/ClaimSubclaims.tsx [hunk 16]
```diff
@@ -170,70 +180,73 @@ const Subclaim: React.FC<{
...
   useEffect(() => {
-    if (subclaims.length === 0 || selectedSubclaim !== null) {
+    if (sortedSubclaims.length === 0 || selectedSubclaim !== null) {
       return
     }
     setSelectedSubclaim(sortedSubclaims[0])
-  }, [sortedSubclaims])
+  }, [sortedSubclaims.length])
+
+  if (!(selectedSubclaim && memberId && claimId)) return null
```
Comment (vonElfvin) [inline src/portals/hope/features/claims/claim-details/ClaimSubclaims/ClaimSubclaims.tsx:220]:
```suggestion
  if (!(selectedSubclaim && memberId && claimId && state)) return null
```

File: src/portals/hope/features/claims/claim-details/SubclaimPayments/SubclaimPayments.tsx [hunk 17]
```diff
@@ -52,7 +53,8 @@ export const SubclaimPayments: React.FC<{
   subclaimId: string
   memberId: string
   claimId: string
-}> = ({ subclaimId, memberId, claimId }) => {
+  state?: string
```
Comment (vonElfvin) [inline src/portals/hope/features/claims/claim-details/SubclaimPayments/SubclaimPayments.tsx:56]: Same here 🤓

File: src/portals/hope/features/claims/claim-details/SubclaimRecovery/SettleModal.tsx [hunk 18]
```diff
@@ -0,0 +1,128 @@
...
+    amount: number
+    note: string
+  }>({
+    defaultValues: {
+      amount: +recovery.expectedAmount.amount,
+    },
+  })
+
+  const submitHandler = async ({ amount, note }: FieldValues) => {
+    if (!+amount || !paymentDate) {
```
Comment (vonElfvin) [inline src/portals/hope/features/claims/claim-details/SubclaimRecovery/SettleModal.tsx:68]:
```suggestion
    if (!amount || +amount <= 0 || !paymentDate) {
```

File: src/portals/hope/features/claims/claim-details/SubclaimRecovery/index.tsx [hunk 19]
```diff
@@ -0,0 +1,351 @@
...
+    )
+
+    await toast.promise(
+      addSubclaimRecovery({
+        variables: {
+          subclaimId,
+          request: {
+            amount: {
+              amount: expectedAmount,
+              currency: 'SEK',
```
Comment (vonElfvin) [inline src/portals/hope/features/claims/claim-details/SubclaimRecovery/index.tsx:192]: Let's drill down the currency using the same solution as Maxim did in the other PR

File: src/portals/hope/features/config/constants.ts [hunk 20]
```diff
@@ -173,6 +173,36 @@ export enum TypeOfContract {
...
 
+export enum LiabilityLevel {
+  Zero = 'ZERO',
+  OneFourth = 'ONE_FOURTH',
+  OneThird = 'ONE_THIRD',
+  Half = 'HALF',
+  TwoThirds = 'TWO_THIRDS',
+  ThreeFourths = 'THREE_FOURTHS',
+  Full = 'FULL',
+}
```
Comment (vonElfvin) [inline src/portals/hope/features/config/constants.ts:184]:
```suggestion
export enum LiabilityLevel {
  Zero = 'ZERO',
  OneThird = 'ONE_THIRD',
  Half = 'HALF',
  TwoThirds = 'TWO_THIRDS',
  Full = 'FULL',
}
```

File: src/portals/hope/features/claims/claim-details/ClaimCounterparty/CounterpartyDetails.tsx [hunk 21]
```diff
@@ -0,0 +1,186 @@
...
+  useUpdateCounterpartyInsuranceProviderMutation,
+  useUpdateCounterpartyLiabilityLevelMutation,
+} from 'types/generated/graphql'
+
+gql`
+  query InsuranceProviders {
+    insuranceProviders {
+      id
+    }
+  }
```
Comment (vonElfvin) [inline src/portals/hope/features/claims/claim-details/ClaimCounterparty/CounterpartyDetails.tsx:22]: We have more on the Insurance Providers now, unsure exactly how to use it though 🤔 
```suggestion
  query InsuranceProviders {
    insuranceProviders {
      id
      displayName
      email
      phone
    }
  }
```

**Review by vonElfvin — COMMENTED** [review]
Very nice! 🔥

File: src/portals/hope/features/claims/ClaimOverview.tsx [hunk 22]
```diff
@@ -265,19 +268,19 @@ export const ClaimOverview: React.FC<{
           </CardsWrapper>
           <CardsWrapper>
             <CardNavigation
-              name="claim-subclaims"
+              name="claim-cases"
```
Comment (vonElfvin) [inline src/portals/hope/features/claims/ClaimOverview.tsx:271]: subclaims should stick

File: src/portals/hope/features/claims/claim-details/ClaimCounterparty/CounterpartyDetails.tsx (hunk 2 above)
Comment (falklinus) [reply src/portals/hope/features/claims/claim-details/ClaimCounterparty/CounterpartyDetails.tsx]: Updated. The solution right now is to leave the "generate invoice" button disabled if there is no registrationNumber

**Review by vonElfvin — APPROVED** [review]

**Pull request merged**